File: recursion/python/solutions/foldr.py
"""Solution for foldr."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def foldr(fn, base, lst):
    """
    Right fold: the pattern behind all these functions!

    foldr(fn, base, a -> b -> c) = fn(a, fn(b, fn(c, base)))

    Examples:
      foldr(add, 0, 1 -> 2 -> 3) = 1 + (2 + (3 + 0)) = 6  (sum)
      foldr(mul, 1, 1 -> 2 -> 3) = 1 * (2 * (3 * 1)) = 6  (product)
      foldr(prepend, (), 1 -> 2 -> 3) = 1 -> 2 -> 3       (copy)

    Once you implement this, you can define:
      sum = foldr(add, 0, lst)
      product = foldr(multiply, 1, lst)
      length = foldr(lambda x, acc: 1 + acc, 0, lst)
      map(f) = foldr(lambda x, acc: prepend(f(x), acc), (), lst)
    """
    if is_empty(lst):
        return base

    first, rest = unprepend(lst)

    return fn(first, foldr(fn, base, rest))


def add(a, b):
    return a + b


def multiply(a, b):
    return a * b


if __name__ == '__main__':
    # Test cases
    assert foldr(add, 0, Node.from_list([1, 2, 3])) == 6
    assert foldr(multiply, 1, Node.from_list([1, 2, 3])) == 6
    assert foldr(add, 0, EMPTY_LIST) == 0
    assert foldr(multiply, 1, EMPTY_LIST) == 1

    # Copy list
    assert to_python_list(foldr(prepend, EMPTY_LIST, Node.from_list([1, 2, 3]))) == [1, 2, 3]

    # Length via fold
    assert foldr(lambda x, acc: 1 + acc, 0, Node.from_list([1, 2, 3, 4])) == 4

    # Map via fold
    double = lambda x, acc: prepend(x * 2, acc)
    assert to_python_list(foldr(double, EMPTY_LIST, Node.from_list([1, 2, 3]))) == [2, 4, 6]

    print("All tests passed!")

File: recursion/javascript/solutions/foldr.js
/**
 * Solution for foldr.
 *
 * This IS the template. Every recursive list function can be written as a foldr.
 *
 * foldr(f, z, [a, b, c]) = f(a, f(b, f(c, z)))
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

function foldr(fn, acc, list) {
  if (isEmpty(list)) {
    return acc;
  }

  let [first, rest] = unprepend(list);

  return fn(first, foldr(fn, acc, rest));
}

// Once you have foldr, everything else is a one-liner:
const sum = (list) => foldr((x, acc) => x + acc, 0, list);
const product = (list) => foldr((x, acc) => x * acc, 1, list);
const length = (list) => foldr((x, acc) => 1 + acc, 0, list);
const map = (fn, list) => foldr((x, acc) => prepend(fn(x), acc), EMPTY_LIST, list);
const filter = (pred, list) => foldr((x, acc) => (pred(x) ? prepend(x, acc) : acc), EMPTY_LIST, list);
const every = (pred, list) => foldr((x, acc) => pred(x) && acc, true, list);
const some = (pred, list) => foldr((x, acc) => pred(x) || acc, false, list);

if (require.main === module) {
  // Direct foldr tests
  console.assert(foldr((x, acc) => x + acc, 0, Node.fromArray([1, 2, 3, 4, 5])) === 15);
  console.assert(foldr((x, acc) => x * acc, 1, Node.fromArray([1, 2, 3, 4, 5])) === 120);

  // Derived function tests
  console.assert(sum(Node.fromArray([1, 2, 3])) === 6);
  console.assert(product(Node.fromArray([2, 3, 4])) === 24);
  console.assert(length(Node.fromArray([1, 2, 3, 4])) === 4);
  console.assert(JSON.stringify(toArray(map((x) => x * 2, Node.fromArray([1, 2, 3])))) === '[2,4,6]');
  console.assert(JSON.stringify(toArray(filter((x) => x % 2 === 0, Node.fromArray([1, 2, 3, 4])))) === '[2,4]');
  console.assert(every((x) => x > 0, Node.fromArray([1, 2, 3])) === true);
  console.assert(some((x) => x < 0, Node.fromArray([1, -2, 3])) === true);

  console.log('All tests passed!');
}

module.exports = { foldr };
